modyn/models/rho_loss_twin_model/rho_loss_twin_model.py

```diff
@@ -0,0 +1,71 @@
+import logging
+from typing import Any, Optional
+
+import torch
+from modyn.utils import dynamic_module_import
+from torch import nn
+
+logger = logging.getLogger(__name__)
+
+
+class RHOLOSSTwinModel:
+
+    def __init__(self, model_configuration: dict[str, Any], device: str, amp: bool) -> None:
+        self.model = RHOLOSSTwinModelModyn(model_configuration, device, amp)
+        self.model.to(device)
+
+
+class RHOLOSSTwinModelModyn(nn.Module):
+
+    def __init__(self, model_configuration: dict[str, Any], device: str, amp: bool) -> None:
+        super().__init__()
+        self.device = device
+        model_module = dynamic_module_import("modyn.models")
+        rho_model_class = model_configuration["rho_real_model_class"]
+        rho_model_config = model_configuration["rho_real_model_config"]
+        model_handler = getattr(model_module, rho_model_class)
+        # we only need the inner model, not the wrapper
+        self._models = nn.ModuleList(
+            [
+                model_handler(rho_model_config, device, amp).model,
+                model_handler(rho_model_config, device, amp).model,
+            ]
+        )
+        self._models_seen_ids: list[set[int]] = [set(), set()]
+        self._current_model = 0
+
+    def get_extra_state(self) -> Any:
+        return {
+            "models_seen_ids": self._models_seen_ids,
+        }
+
+    def set_extra_state(self, state: dict) -> None:
+        self._models_seen_ids = state["models_seen_ids"]
+        # the second time we train on this model, we should switch to the other model
+        self._current_model = 1
+
+    def forward(self, data: torch.Tensor, sample_ids: Optional[list[int]] = None) -> torch.Tensor:
+        assert sample_ids is not None
+        if self.training:
```

Comment:
where is self.training coming from? fields should be initialized in the constructor, maybe I don't see it. Usually pylint should complain here that you access a field not initialized in the constructor

Reply:
also, why not make this decision based on self.current_model?

Reply:
but actually I do not understand yet how this is updated. can you maybe add a comment here explaining this? there is nothing in the trainer server that sets this, how does the model know?

Reply:
I will add it. It is an attribute in `nn.Module` that's updated everytime `.train()` or `.eval()` is called.
See https://discuss.pytorch.org/t/what-does-model-training-in-the-transfer-learning-tutorial-mean/82182